src/hotspot/share/nmt/nmtTreap.hpp

```diff
@@ -0,0 +1,282 @@
+      head->_right = p.left;
+      return nd_pair{head, p.right};
+    } else {
+      nd_pair p = split(head->_left, key, mode);
+      head->_left = p.right;
+      return nd_pair{p.left, head};
+    }
+  }
+
+  // Invariant: left is a treap whose keys are LEQ to the keys in right.
+  static Nd* merge(Nd* left, Nd* right) {
+    if (left == nullptr) return right;
+    if (right == nullptr) return left;
+
+    if (left->_priority > right->_priority) {
+      // We need
+      //      LEFT
+      //         |
+      //         RIGHT
+      // For the invariant re: priorities to hold.
+      left->_right = merge(left->_right, right);
+      return left;
+    } else {
+      // We need
+      //         RIGHT
+      //         |
+      //      LEFT
+      // For the invariant re: priorities to hold.
+      right->_left = merge(left, right->_left);
+      return right;
+    }
+  }
+
+public:
+  TreapNode(const K& k, const V& v, uint64_t p)
+  : _priority(p), _key(k), _value(v), _left(nullptr), _right(nullptr) {
+  }
+
+  const K& key() const {
+    return _key;
+  }
+  V& val() {
+    return _value;
+  }
+
+  Nd* left() const {
+    return _left;
+  }
+  Nd* right() const {
+    return _right;
+  }
+
+  static Nd* find(Nd* node, const K& k) {
+    if (node == nullptr) {
+      return nullptr;
+    }
+    if (CMP(node->_key, k) == 0) { // EQ
+      return node;
+    }
+
+    if (CMP(node->_key, k) <= 0) { // LEQ
+      return find(node->_left, k);
+    } else {
+      return find(node->_right, k);
+    }
+  }
+
+  template<typename MakeNode>
+  static Nd* upsert(Nd* head, const K& k, const V& v, MakeNode make_node) {
+    // (LEQ_k, GT_k)
+    nd_pair split = Nd::split(head, k);
+    Nd* found = find(split.left, k);
+    if (found != nullptr) {
+      // Already exists, update value.
+      found->_value = v;
+      return merge(split.left, split.right);
+    }
+    // Doesn't exist, make node
+    Nd* node = make_node(k, v);
+    // merge(merge(LEQ_k, EQ_k), GT_k)
+    return merge(merge(split.left, node), split.right);
+  }
+
+  template<typename Free>
+  static Nd* remove(Nd *head, const K& k, Free free) {
+    // (LEQ_k, GT_k)
+    nd_pair fst_split = split(head, k, LEQ);
+    // (LT_k, GEQ_k) == (LT_k, EQ_k) since it's from LEQ_k and keys are unique.
+    nd_pair snd_split = split(fst_split.left, k, LT);
+
+    if (snd_split.right != nullptr) {
+      // The key k existed, we delete it.
+      free(snd_split.right);
+    }
+    // Merge together everything
+    return merge(snd_split.left, fst_split.right);
+  }
+
+  // Delete all nodes.
+  template<typename Free>
+  static Nd* remove_all(Nd* tree, Free free) {
+    GrowableArrayCHeap<Nd*, mtNMT> to_delete;
+    to_delete.push(tree);
+
+    while (!to_delete.is_empty()) {
+      Nd* head = to_delete.pop();
+      if (head == nullptr) continue;
+      to_delete.push(head->_left);
+      to_delete.push(head->_right);
+      free(head);
+    }
+    return nullptr;
+  }
+};
+
```

Comment:
I dislike having all tree logic in a tree class that is already bound to a specific allocation pattern. I may want to reuse the tree with different allocators, e.g. to avoid malloc. Another reason to make the allocator part of the template parameter of a tree.